### PR TITLE
Out-of-band steering data support.

### PR DESCRIPTION
**Pull request by vaas-krish**

### CODE REVIEW

File: examples/drivers/windows/otApi/otApi.cpp [hunk 1]
```diff
@@ -1347,7 +1348,7 @@ otThreadDiscover(
         aInstance->InterfaceGuid, aCallback, aCallbackContext
         );
 
-    PackedBuffer4<GUID,uint32_t,uint16_t, bool> Buffer(aInstance->InterfaceGuid, aScanChannels, aPanid, aJoiner);
+    PackedBuffer4<GUID,uint32_t,uint16_t, bool, bool> Buffer(aInstance->InterfaceGuid, aScanChannels, aPanid, aJoiner, aEnableEui64Filtering);
```
Comment (nibanks) [inline examples/drivers/windows/otApi/otApi.cpp]: `PackedBuffer5` needs to be used.

File: examples/drivers/windows/otLwf/iocontrol.c [hunk 2]
```diff
@@ -915,6 +915,7 @@ otLwfIoCtl_otDiscover(
                 aScanChannels, 
                 aPanid,
                 aJoiner,
+                aEnableEui64Filtering,
```
Comment (nibanks) [inline examples/drivers/windows/otLwf/iocontrol.c]: This variable isn't declared.

File: examples/drivers/windows/otApi/otApi.cpp (hunk 1 above)
Comment (vaas-krish) [inline examples/drivers/windows/otApi/otApi.cpp]: fixed.

File: examples/drivers/windows/otLwf/iocontrol.c (hunk 2 above)
Comment (vaas-krish) [inline examples/drivers/windows/otLwf/iocontrol.c]: fixed.

File: src/core/meshcop/tlvs.hpp [hunk 3]
```diff
@@ -596,6 +597,49 @@ class SteeringDataTlv: public Tlv
     }
 
     /**
+     * Ths method indicates whether or not the SteeringData is all zeros.
+     *
+     * @retval TRUE   If the SteeringData is all zeros.
+     * @retval FALSE  If the SteeringData isn't all zeros.
+     *
+     */
+    bool IsCleared(void) {
```
Comment (vaas-krish) [inline src/core/meshcop/tlvs.hpp]: done

**Comment by jwhui** [comment]
It looks like binaries were accidentally included in this PR.

File: src/core/meshcop/meshcop_tlvs.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
+/*
```
Comment (jwhui) [inline src/core/meshcop/meshcop_tlvs.cpp]: We have `meshcop/meshcop_tlvs.cpp` and `meshcop/tlvs.hpp`.  Prefer to have the filenames match between the `.cpp` and `.hpp` files.

File: src/core/meshcop/tlvs.hpp [hunk 5]
```diff
@@ -630,6 +631,23 @@ class SteeringDataTlv: public Tlv
...
     void SetBit(uint8_t aBit) { mSteeringData[GetLength() - 1 - (aBit / 8)] |= 1 << (aBit % 8); }
 
+    /**
+     * Ths method indicates whether or not the SteeringData is all zeros.
+     *
+     * @retval TRUE   If the SteeringData is all zeros.
+     * @retval FALSE  If the SteeringData isn't all zeros.
+     *
+     */
+    bool IsCleared(void);
```
Comment (jwhui) [inline src/core/meshcop/tlvs.hpp]: Declare as `const`?

File: src/core/thread/mle_router.cpp [hunk 6]
```diff
@@ -2464,6 +2464,39 @@ ThreadError MleRouter::HandleNetworkDataUpdateRouter(void)
...
+    }
+    else
+    {
+        // Set bloom filter with the extended address passed in
+        mSteeringData.ComputeBloomFilter(aExtAddress);
+    }
+
+    return error;
+}
+#endif
```
Comment (jwhui) [inline src/core/thread/mle_router.cpp]: Add trailing comment to indicate matching `#if`.

File: src/core/thread/mle_router.cpp [hunk 7]
```diff
@@ -2512,7 +2545,11 @@ ThreadError MleRouter::HandleDiscoveryRequest(const Message &aMessage, const Ip6
 
             if (discoveryRequest.IsJoiner())
             {
+#if OPENTHREAD_CONFIG_ENABLE_STEERING_DATA_SET_OOB
+                VerifyOrExit(!mSteeringData.IsCleared());
```
Comment (jwhui) [inline src/core/thread/mle_router.cpp]: If OOB steering data is not set, should it fall back to network data?

Comment (vaas-krish) [inline src/core/thread/mle_router.cpp]: @jwhui, addressed all the comments in the latest commit.

File: examples/drivers/windows/otApi/otApi.cpp [hunk 8]
```diff
@@ -1347,7 +1348,7 @@ otThreadDiscover(
         aInstance->InterfaceGuid, aCallback, aCallbackContext
         );
 
-    PackedBuffer4<GUID,uint32_t,uint16_t, bool> Buffer(aInstance->InterfaceGuid, aScanChannels, aPanid, aJoiner);
+    PackedBuffer5<GUID,uint32_t,uint16_t, bool, bool> Buffer(aInstance->InterfaceGuid, aScanChannels, aPanid, aJoiner, aEnableEui64Filtering);
```
Comment (nibanks) [inline examples/drivers/windows/otApi/otApi.cpp]: I was just looking at this more closely, and I don't think we should use `bool` here (for either of the last two types), and instead use `uint8_t` to explicitly set the size. `iocontrol.c` would also need to be updated.

Comment (vaas-krish) [inline examples/drivers/windows/otApi/otApi.cpp]: Done.

**Pull request merged**